cmd/snap-update-ns/utils.go

```diff
@@ -667,3 +667,92 @@ func createWritableMimic(dir, neededBy string, sec *Secure) ([]*Change, error) {
 	}
 	return changes, nil
 }
+
+// PathIterator traverses through segments (directories and files) of some
+// path. The filesystem is never consulted, traversal is done purely in memory.
+//
+// The iterator is useful in implementing secure traversal of absolute paths
+// using the common idiom of opening the root directory followed by a chain of
+// openat calls.
+//
+// A simple example on how to use the iterator:
+// ```
+// pi := NewPathIterator(path)
+// for pi.Next() {
+//    // pi.PathSoFar() or pi.Segment()
+// }
+// ```
+type PathIterator struct {
+	path        string
+	left, right int
+}
+
+// NewPathIterator returns an iterator for traversing the given path.
+// The path is passed through filepath.Clean automatically.
+func NewPathIterator(path string) *PathIterator {
+	return &PathIterator{path: filepath.Clean(path)}
+}
+
+// Path returns the path being traversed.
+func (pi *PathIterator) Path() string {
+	return pi.path
+}
+
+// Segment returns the name of the current path segment.
+//
+// Segment may return the empty string to denote the root directory.
```

Comment:
AFAICT `CleanSegment()` will return empty string for `/`, this will will just return `/`.

Reply:
Yes, I fixed this, I will post a follow up shortly (I have a full branch using this so some shortcomings came out)